FWCore/Integration/test/TestInterProcessRandomProd.cc

```diff
@@ -0,0 +1,250 @@
+        : id_{id},
+          channel_("testProd", id_),
+          readBuffer_{channel_.sharedMemoryName(), channel_.fromWorkerBufferIndex()},
+          writeBuffer_{std::string("Rand") + channel_.sharedMemoryName(), channel_.toWorkerBufferIndex()},
+          deserializer_{readBuffer_},
+          bl_deserializer_{readBuffer_},
+          randSerializer_{writeBuffer_} {
+      //make sure output is flushed before popen does any writing
+      fflush(stdout);
+      fflush(stderr);
+
+      channel_.setupWorker([&]() {
+        using namespace std::string_literals;
+        std::cout << id_ << " starting external process" << std::endl;
+        pipe_ = popen(("cmsTestInterProcessRandom "s + channel_.sharedMemoryName() + " " + channel_.uniqueID()).c_str(),
+                      "w");
+
+        if (NULL == pipe_) {
+          abort();
+        }
+
+        {
+          auto nlines = std::to_string(std::count(iConfig.begin(), iConfig.end(), '\n'));
+          auto result = fwrite(nlines.data(), sizeof(char), nlines.size(), pipe_);
+          assert(result = nlines.size());
+          result = fwrite(iConfig.data(), sizeof(char), iConfig.size(), pipe_);
+          assert(result == iConfig.size());
+          fflush(pipe_);
+        }
+      });
+    }
+
+    template <typename SERIAL>
+    auto doTransition(SERIAL& iDeserializer, edm::Transition iTrans, unsigned long long iTransitionID)
+        -> decltype(iDeserializer.deserialize()) {
+      decltype(iDeserializer.deserialize()) value;
+      if (not channel_.doTransition(
+              [&value, this]() {
+                value = deserializer_.deserialize();
+                std::cout << id_ << " from shared memory " << value.first.value << std::endl;
+              },
+              iTrans,
+              iTransitionID)) {
+        std::cout << id_ << " FAILED waiting for external process" << std::endl;
+        externalFailed_ = true;
+        throw cms::Exception("ExternalFailed");
+      }
+      return value;
+    }
+    edmtest::IntProduct produce(unsigned long long iTransitionID, edm::StreamID iStream) {
+      edm::Service<edm::RandomNumberGenerator> gen;
+      auto& engine = gen->getEngine(iStream);
+      edm::RandomNumberGeneratorState state{engine.put(), engine.getSeed()};
+      randSerializer_.serialize(state);
+      auto v = doTransition(deserializer_, edm::Transition::Event, iTransitionID);
+      engine.setSeed(v.second.seed_, 0);
+      engine.get(v.second.state_);
+      return v.first;
+    }
+
+    ReturnedType beginLumiProduce(edm::RandomNumberGeneratorState const& iState,
+                                  unsigned long long iTransitionID,
+                                  edm::LuminosityBlockIndex iLumi) {
+      edm::Service<edm::RandomNumberGenerator> gen;
+      //NOTE: root serialize requires a `void*` not a `void const*` even though it doesn't modify the object
+      randSerializer_.serialize(const_cast<edm::RandomNumberGeneratorState&>(iState));
```

Comment:
Can we change `ROOTSerializer` to take `void const*`? (in a later PR)

Reply:
We can. The underlying problem is ROOT's code which uses the 'void*' instead of the 'void const*'.

Reply:
> We can. The underlying problem is ROOT's code which uses the 'void*' instead of the 'void const*'.

I guessed that would be the case. At least the `const_cast` would then be isolated and hidden from the rest...